### PR TITLE
Refactor/settings export

### LINKED ISSUE #173: Importing settings with a deleted account set as the initial account results in login page being shown

### LINKED ISSUE #496: Settings Export/Import broken
The SettingsExporter is broken in multiple ways. This even dates back to before #480

### PR DESCRIPTION
**Pull request by Leptopoda**
Closes #496
Closes #173 

Compleatly re implements the `SettingsExportHelper`.

---

On a sidenote: Why are we base64 encoding the exported data? wouldn't it be sufficient enough to safe the json directly?

### CODE REVIEW

**Comment by Leptopoda** [comment]
Ci will be pasing with https://github.com/provokateurin/nextcloud-neon/pull/503

**Comment by Leptopoda** [comment]
I somehow dislike this :(
I think we should add an `Exportable` interface or something like that and let the AccountsBloc or whatever else we'll have define what they want to export. The downside is that the exporting/importing logic wouldn't be in one place anymore.

**Comment by Leptopoda** [comment]
I somehow dislike this :(
I think we should add an `Exportable` interface or something like that and let the AccountsBloc or whatever else we'll have define what they want to export. The downside is that the exporting/importing logic wouldn't be in one place anymore.

**Comment by provokateurin** [comment]
The base64 encoding makes no sense, no idea why it is there. I like the exportable interface :+1:

File: packages/neon/neon/lib/src/settings/models/exportable.dart [hunk 1]
```diff
@@ -0,0 +1,35 @@
+import 'dart:async';
+
+import 'package:collection/collection.dart';
+import 'package:neon/src/settings/models/option.dart';
+
+/// Exportable data interface.
+abstract interface class Exportable {
+  /// Exports this into a json map entry.
```
Comment (provokateurin) [inline packages/neon/neon/lib/src/settings/models/exportable.dart:8]:
```suggestion
  /// Exports into a json map entry.
```

File: packages/neon/neon/lib/src/settings/models/exportable.dart [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  MapEntry<String, Object?> export();
+
+  /// Imports [data] from an export.
+  FutureOr<void> import(final Map<String, Object?> data);
+}
+
+/// Serialization helpers for a collection of [Option]s.
+extension SerializeOptions on Iterable<Option<dynamic>> {
+  /// Serializes this into  an [Iterable<JsonEntry>].
```
Comment (provokateurin) [inline packages/neon/neon/lib/src/settings/models/exportable.dart:17]:
```suggestion
  /// Serializes into an [Iterable<JsonEntry>].
```

File: packages/neon/neon/lib/src/settings/models/nextcloud_app_options.dart [hunk 3]
```diff
@@ -1,23 +1,66 @@
...
+import 'package:neon/src/settings/models/exportable.dart';
 import 'package:neon/src/settings/models/option.dart';
 import 'package:neon/src/settings/models/options_category.dart';
 import 'package:neon/src/settings/models/storage.dart';
 
-abstract class NextcloudAppOptions {
-  NextcloudAppOptions(this.storage);
+/// Collection of [Option]s.
+abstract class OptionsCollection implements Exportable {
+  /// Creates a new OptionsCollection.
```
Comment (provokateurin) [inline packages/neon/neon/lib/src/settings/models/nextcloud_app_options.dart:9]: Not really helpful, I usually just ignore constructors

File: packages/neon/neon/lib/src/settings/models/nextcloud_app_options.dart [hunk 4]
```diff
@@ -1,23 +1,66 @@
...
+  @override
+  void import(final Map<String, Object?> data) {
+    final value = data[storage.id];
+    if (value != null) {
+      options.deserialize(value as Map<String, Object?>);
+    }
+  }
+}
+
+/// OpptionsCollection for a neon app.
```
Comment (provokateurin) [inline packages/neon/neon/lib/src/settings/models/nextcloud_app_options.dart:56]:
```suggestion
/// OpptionsCollection for an [AppImplementation].
```

Is this better?

File: packages/neon/neon/lib/src/settings/models/nextcloud_app_options.dart [hunk 5]
```diff
@@ -1,23 +1,66 @@
...
+abstract class NextcloudAppOptions extends OptionsCollection {
+  /// Creates a new NextcloudAppOptions.
```
Comment (provokateurin) [inline packages/neon/neon/lib/src/settings/models/nextcloud_app_options.dart:58]: Also not helpful imo

File: packages/neon/neon/lib/src/settings/models/option.dart [hunk 6]
```diff
@@ -83,6 +83,15 @@ abstract class Option<T> extends ChangeNotifier implements ValueListenable<T> {
     value = defaultValue;
   }
 
+  /// loads [data] into [value] by calling [deserialize] on it.
```
Comment (provokateurin) [inline packages/neon/neon/lib/src/settings/models/option.dart:86]:
```suggestion
  /// Loads [data] into [value] by calling [deserialize] on it.
```

**Review by provokateurin — CHANGES_REQUESTED** [review]

File: packages/neon/neon/lib/src/settings/models/nextcloud_app_options.dart (hunk 3 above)
Comment (Leptopoda) [reply packages/neon/neon/lib/src/settings/models/nextcloud_app_options.dart:9]: Should I remove them?

File: packages/neon/neon/lib/src/settings/models/exportable.dart (hunk 1 above)
Comment (Leptopoda) [reply packages/neon/neon/lib/src/settings/models/exportable.dart:8]: Why without the this?

File: packages/neon/neon/lib/src/settings/models/nextcloud_app_options.dart (hunk 3 above)
Comment (provokateurin) [reply packages/neon/neon/lib/src/settings/models/nextcloud_app_options.dart:9]: Yes please

File: packages/neon/neon/lib/src/settings/models/nextcloud_app_options.dart (hunk 4 above)
Comment (Leptopoda) [reply packages/neon/neon/lib/src/settings/models/nextcloud_app_options.dart:56]: The thing is that `AppImplementation` is not visible in that scope and just importing it for the reference in the comment doesn't feel right

Comment (provokateurin) [reply packages/neon/neon/lib/src/settings/models/nextcloud_app_options.dart:56]: Alright

**Review by provokateurin — CHANGES_REQUESTED** [review]
Importing settings does not work correctly for apps and accounts. The export definitely includes the correct values.

**Comment by Leptopoda** [comment]
I also took the chance and cleaned up our AppStorage :)

**Comment by provokateurin** [comment]
Now apps and accounts work, but the global options like theme all don't work :grimacing:

**Comment by provokateurin** [comment]
I think unit tests for this would be great otherwise it will probably break very quickly again :/

**Comment by Leptopoda** [comment]
Unit tests are hard especially around our blocs :(

**Comment by Leptopoda** [comment]
My bad The initial version was hard to unit test but the current code should make it fairly simple.

I'll add a few test later.

**Comment by Leptopoda** [comment]
@provokateurin  I've added some unit tests and also updated the `Option` tests to get 100% coverage again.

I didn't test the specifics of the options though as they are to entangled with the blocs for now.

**Comment by provokateurin** [comment]
Please rebase then I can properly review :)

**Review by provokateurin — CHANGES_REQUESTED** [review]
Importing account settings now works, but app settings are still broken.

**Comment by Leptopoda** [comment]
Ok I just rebased onto main (aka #688 as main would be broken) and I can't reproduce the issue.
I tested it with every app, GlobalOptions and AccountOptions (It would be very interesting as AFICT the export part has 100% coverage and I'm fairly sure the tests are not false).

Can you please verify that you had the latest version of the PR checked out?
Can you also provide further information of what exact settings are broken or even provide the exported json.

Here is the export I successfully verified as working:
[nextcloud-neon-settings-1693437863.json.txt](https://github.com/nextcloud/neon/files/12481254/nextcloud-neon-settings-1693437863.json.txt)

**Review by provokateurin — APPROVED** [review]
You are right, it was working correctly. I just so happened to find https://github.com/nextcloud/neon/pull/689 by accident and thought the import wasn't working for that option. Somewhat lucky that I found it.

**Pull request merged**